File: libs/python/financequote.py
"""
FinanceQuote Python Client

A simple client library for the FinanceQuote API.
"""

import requests
from typing import Optional, Dict, List, Any


class FinanceQuoteClient:
    """Client for FinanceQuote API."""
    
    def __init__(self, base_url: str, api_key: Optional[str] = None):
        """
        Initialize the client.
        
        Args:
            base_url: The API base URL (e.g., "http://localhost:3001")
            api_key: Optional API key for authentication
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.session = requests.Session()
    
    def _get_headers(self) -> Dict[str, str]:
        """Get request headers including auth."""
        headers = {'Content-Type': 'application/json'}
        if self.api_key:
            headers['Authorization'] = f'Bearer {self.api_key}'
        return headers
    
    def get_quote(self, symbol: str, method: str = 'yahoojson', 
                  currency: Optional[str] = None) -> Dict[str, Any]:
        """
        Get a single stock quote.
        
        Args:
            symbol: Stock ticker symbol (e.g., "AAPL")
            method: Quote method (default: "yahoojson")
            currency: Optional currency for conversion
        
        Returns:
            Quote data dictionary
        """
        result = self.get_quotes([symbol], method, currency)
        if symbol in result:
            return result[symbol]
        raise ValueError(f"No quote found for {symbol}")
    
    def get_quotes(self, symbols: List[str], method: str = 'yahoojson',
                   currency: Optional[str] = None) -> Dict[str, Dict[str, Any]]:
        """
        Get quotes for multiple symbols.
        
        Args:
            symbols: List of stock ticker symbols (e.g., ["AAPL", "GOOGL"])
            method: Quote method (default: "yahoojson")
            currency: Optional currency for conversion
        
        Returns:
            Dictionary mapping symbol to quote data
        """
        if not symbols:
            raise ValueError("At least one symbol required")
        
        url = f"{self.base_url}/api/v1/quote/{','.join(symbols)}"
        
        params = {}
        if method:
            params['method'] = method
        if currency:
            params['currency'] = currency
        
        response = self.session.get(url, headers=self._get_headers(), params=params)
        response.raise_for_status()
        
        data = response.json()
        
        if data.get('status') != 'success':
            error_info = data.get('error', {})
            raise ValueError(f"API error: {error_info.get('message', 'Unknown error')}")
        
        # Filter out non-quote data
        result = {k: v for k, v in data.get('data', {}).items() 
                  if k != 'methods' and isinstance(v, dict)}
        
        return result
    
    def get_methods(self) -> List[str]:
        """
        Get all available quote methods.
        
        Returns:
            List of available methods
        """
        url = f"{self.base_url}/api/v1/methods"
        
        response = self.session.get(url, headers=self._get_headers())
        response.raise_for_status()
        
        data = response.json()
        return data.get('data', {}).get('methods', [])
    
    def get_currency(self, from_currency: str, to_currency: str) -> float:
        """
        Get currency conversion rate.
        
        Args:
            from_currency: Source currency code (e.g., "USD")
            to_currency: Target currency code (e.g., "EUR")
        
        Returns:
            Conversion rate
        """
        url = f"{self.base_url}/api/v1/currency/{from_currency}/{to_currency}"
        
        response = self.session.get(url, headers=self._get_headers())
        response.raise_for_status()
        
        data = response.json()
        
        if data.get('status') != 'success':
            error_info = data.get('error', {})
            raise ValueError(f"Currency error: {error_info.get('message', 'Unknown error')}")
        
        # The API returns {status: "success", data: {from, to, rate}}
        rate_data = data.get('data', {})
        rate = rate_data.get('rate', 0)
        
        return float(rate)
    
    def health_check(self) -> bool:
        """
        Check if the API is healthy.
        
        Returns:
            True if healthy, False otherwise
        """
        try:
            url = f"{self.base_url}/api/v1/health"
            response = self.session.get(url)
            return response.status_code == 200
        except requests.RequestException:
            return False


# Convenience functions

def quote(symbol: str, url: str = "http://localhost:3001", 
          api_key: Optional[str] = None, **kwargs) -> Dict[str, Any]:
    """
    Get a single stock quote.
    
    Example:
        >>> quote("AAPL", "http://localhost:3001")
        {'symbol': 'AAPL', 'last': 260.48, ...}
    """
    client = FinanceQuoteClient(url, api_key)
    return client.get_quote(symbol, **kwargs)


def quotes(symbols: List[str], url: str = "http://localhost:3001",
           api_key: Optional[str] = None, **kwargs) -> Dict[str, Dict[str, Any]]:
    """
    Get multiple stock quotes.
    
    Example:
        >>> quotes(["AAPL", "GOOGL"], "http://localhost:3001")
        {'AAPL': {...}, 'GOOGL': {...}}
    """
    client = FinanceQuoteClient(url, api_key)
    return client.get_quotes(symbols, **kwargs)


def currency(from_curr: str, to_curr: str, 
             url: str = "http://localhost:3001",
             api_key: Optional[str] = None) -> float:
    """
    Get currency conversion rate.
    
    Example:
        >>> currency("USD", "EUR", "http://localhost:3001")
        0.92
    """
    client = FinanceQuoteClient(url, api_key)
    return client.get_currency(from_curr, to_curr)


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = FinanceQuoteClient("http://localhost:3001")
    
    # Get single quote
    apple = client.get_quote("AAPL")
    print(f"Apple: ${apple['last']}")
    
    # Get multiple quotes
    quotes = client.get_quotes(["AAPL", "GOOGL", "MSFT"])
    for symbol, data in quotes.items():
        print(f"{symbol}: ${data.get('last', 'N/A')}")
    
    # Get available methods
    methods = client.get_methods()
    print(f"Available methods: {len(methods)}")
    
    # Check health
    print(f"API healthy: {client.health_check()}")

File: libs/node/financequote.js
/**
 * FinanceQuote Node.js Client
 * A simple client library for the FinanceQuote API.
 */

const http = require('http');
const https = require('https');

class FinanceQuoteClient {
  /**
   * Create a new FinanceQuote API client.
   * @param {string} baseURL - The API base URL (e.g., "http://localhost:3001")
   * @param {string} [apiKey] - Optional API key for authentication
   */
  constructor(baseURL, apiKey = null) {
    this.baseURL = baseURL.replace(/\/$/, '');
    this.apiKey = apiKey;
  }

  _getHeaders() {
    const headers = { 'Content-Type': 'application/json' };
    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }
    return headers;
  }

  _makeRequest(path) {
    return new Promise((resolve, reject) => {
      const url = new URL(path, this.baseURL);
      const client = url.protocol === 'https:' ? https : http;

      const options = {
        hostname: url.hostname,
        port: url.port || (url.protocol === 'https:' ? 443 : 80),
        path: url.pathname + url.search,
        method: 'GET',
        headers: this._getHeaders()
      };

      const req = client.request(options, (res) => {
        let data = '';
        res.on('data', chunk => data += chunk);
        res.on('end', () => {
          try {
            const json = JSON.parse(data);
            if (json.status === 'success') {
              resolve(json.data);
            } else {
              reject(new Error(json.error?.message || 'API error'));
            }
          } catch (e) {
            reject(e);
          }
        });
      });

      req.on('error', reject);
      req.end();
    });
  }

  /**
   * Get a single stock quote.
   * @param {string} symbol - Stock ticker symbol (e.g., "AAPL")
   * @param {object} [params] - Optional parameters
   * @param {string} [params.method] - Quote method (default: "yahoojson")
   * @param {string} [params.currency] - Currency for conversion
   * @returns {Promise<object>} Quote data
   */
  async getQuote(symbol, params = {}) {
    const symbols = await this.getQuotes([symbol], params);
    if (symbols[symbol]) {
      return symbols[symbol];
    }
    throw new Error(`No quote found for ${symbol}`);
  }

  /**
   * Get quotes for multiple symbols.
   * @param {string[]} symbols - Array of stock ticker symbols
   * @param {object} [params] - Optional parameters
   * @param {string} [params.method] - Quote method (default: "yahoojson")
   * @param {string} [params.currency] - Currency for conversion
   * @returns {Promise<object>} Dictionary of quotes
   */
  async getQuotes(symbols, params = {}) {
    if (!symbols || symbols.length === 0) {
      throw new Error('At least one symbol required');
    }

    const query = new URLSearchParams();
    if (params.method) query.set('method', params.method);
    if (params.currency) query.set('currency', params.currency);

    const path = `/api/v1/quote/${symbols.join(',')}${query.toString() ? '?' + query.toString() : ''}`;
    const data = await this._makeRequest(path);

    // Filter out non-quote data
    const result = {};
    for (const [key, value] of Object.entries(data)) {
      if (key !== 'methods' && typeof value === 'object') {
        result[key] = value;
      }
    }
    return result;
  }

  /**
   * Get all available quote methods.
   * @returns {Promise<string[]>} List of available methods
   */
  async getMethods() {
    const data = await this._makeRequest('/api/v1/methods');
    return data.methods || [];
  }

  /**
   * Get currency conversion rate.
   * @param {string} fromCurrency - Source currency code (e.g., "USD")
   * @param {string} toCurrency - Target currency code (e.g., "EUR")
   * @returns {Promise<number>} Conversion rate
   */
  async getCurrency(fromCurrency, toCurrency) {
    // API returns {status: "success", data: {from, to, rate}}
    const data = await this._makeRequest(`/api/v1/currency/${fromCurrency}/${toCurrency}`);
    if (data && data.rate !== undefined) {
      return parseFloat(data.rate);
    }
    throw new Error('Currency rate not available');
  }

  /**
   * Check if the API is healthy.
   * @returns {Promise<boolean>} True if healthy
   */
  async healthCheck() {
    try {
      const data = await this._makeRequest('/api/v1/health');
      return data && data.service === 'FinanceQuote API';
    } catch {
      return false;
    }
  }
}

// Convenience functions
async function quote(symbol, url = 'http://localhost:3001', apiKey = null, params = {}) {
  const client = new FinanceQuoteClient(url, apiKey);
  return client.getQuote(symbol, params);
}

async function quotes(symbols, url = 'http://localhost:3001', apiKey = null, params = {}) {
  const client = new FinanceQuoteClient(url, apiKey);
  return client.getQuotes(symbols, params);
}

async function currency(from, to, url = 'http://localhost:3001', apiKey = null) {
  const client = new FinanceQuoteClient(url, apiKey);
  return client.getCurrency(from, to);
}

// Export for CommonJS and ES modules
module.exports = {
  FinanceQuoteClient,
  quote,
  quotes,
  currency
};

// Example usage
if (require.main === module) {
  (async () => {
    const client = new FinanceQuoteClient('http://localhost:3001');

    try {
      // Get single quote
      const apple = await client.getQuote('AAPL');
      console.log('Apple:', apple.last);

      // Get multiple quotes
      const quotes = await client.getQuotes(['AAPL', 'GOOGL', 'MSFT']);
      for (const [symbol, data] of Object.entries(quotes)) {
        console.log(`${symbol}: $${data.last || 'N/A'}`);
      }

      // Get methods
      const methods = await client.getMethods();
      console.log(`Available methods: ${methods.length}`);

      // Health check
      console.log(`API healthy: ${await client.healthCheck()}`);
    } catch (err) {
      console.error('Error:', err.message);
    }
  })();
}